File: src/data/courses/dsa/two-pointers-sliding-window/sliding-window-variable-shortest/minimum-window-substring/code/cpp/minimum_window_substring_sol1.cpp
class Solution {
public:
    string minWindow(string s, string t) {
        string result; int minLen = INT_MAX;
        for (int i = 0; i < s.size(); i++) {
            unordered_map<char,int> need;
            for (char c : t) need[c]++;
            for (int j = i; j < s.size(); j++) {
                if (need.count(s[j])) need[s[j]]--;
                bool valid = true;
                for (auto& [c,cnt] : need) if (cnt > 0) { valid = false; break; }
                if (valid && j-i+1 < minLen) { minLen = j-i+1; result = s.substr(i, minLen); break; }
            }
        }
        return result;
    }
};

File: src/data/courses/dsa/two-pointers-sliding-window/sliding-window-variable-shortest/minimum-window-substring/code/python/minimum_window_substring_sol1.py
class Solution:
    def minWindow(self, s, t):
        from collections import Counter
        result, min_len = '', float('inf')
        for i in range(len(s)):
            need = Counter(t)
            for j in range(i, len(s)):
                if s[j] in need: need[s[j]] -= 1
                if all(v <= 0 for v in need.values()):
                    if j-i+1 < min_len: min_len = j-i+1; result = s[i:j+1]
                    break
        return result